### PR TITLE
Support for parsing connection strings in ODBC and URL formats

### PR DESCRIPTION
**Pull request by arsing**
Ref #169

### CODE REVIEW

File: tds.go [hunk 1]
```diff
@@ -691,9 +693,240 @@ func splitConnectionString(dsn string) (res map[string]string) {
 	return res
 }
 
+// Splits a URL in the ODBC format
+func splitConnectionStringOdbc(dsn string) (map[string]string, error) {
+	res := map[string]string{}
+
+	type parserState int
+	const (
```
Comment (kardianos) [inline tds.go]: What you want is:
```
const (
    paserStateBetweenPairs parserState = iota // comment a
    parserStateKey                        // comment b
    parserStateBeginValue                  // comment c
    ...
)
```

File: tds.go [hunk 2]
```diff
@@ -691,9 +693,240 @@ func splitConnectionString(dsn string) (res map[string]string) {
...
+
+	var state = parserStateBetweenPairs
+
+	var key string
+	var value string
+
+	for i, c := range dsn {
+		switch state {
+		case parserStateBetweenPairs:
+			if c == '=' {
```
Comment (kardianos) [inline tds.go]: I'm okay with these if/else statements. I would prefer them to be switch statements:
```
switch {
case c == '=':
    return res, err
case !unicode.IsSpace(c) && c != ';':
    ...
}
```

File: tds.go [hunk 3]
```diff
@@ -691,9 +693,240 @@ func splitConnectionString(dsn string) (res map[string]string) {
...
+				state = parserStateBetweenPairs
+			} else if unicode.IsSpace(c) {
+				// Ignore whitespace
+			} else {
+				return res, fmt.Errorf("Unexpected character %c at index %d. Expected semi-colon or whitespace.", c, i)
+			}
+		}
+	}
+
+	switch state {
```
Comment (kardianos) [inline tds.go]: Please get rid of extra space:
```
case parserStateBetweenPairs: // Okay.
case parserStateKey: // Unfinished key...
    key = normalized..
case parserStateBeginValue: // EmptyValue.
   ...
}
```

File: tds.go [hunk 4]
```diff
@@ -691,9 +693,240 @@ func splitConnectionString(dsn string) (res map[string]string) {
...
+			// End of braced value
+			res[key] = value
+			key = ""
+			value = ""
+			state = parserStateBetweenPairs
+
+			// This character is the first character past the end,
+			// so it needs to be parsed as if this is the  parserStateEndValue state.
+			state = parserStateEndValue
+			fallthrough
```
Comment (kardianos) [inline tds.go]: Remove `fallthrough` please.

**Comment by kardianos** [comment]
Also, please add appropriate docs for these new formats to the README.

File: tds.go (hunk 1 above)
Comment (arsing) [inline tds.go]: I'll leave the comments above the fields since one of them is multi-line.

Comment (arsing) [inline tds.go]: Actually the `parserState` is also required, since `iota` is just `int.

File: tds.go (hunk 4 above)
Comment (arsing) [inline tds.go]: Sure, though it seems cleaner to not duplicate the code...

File: tds.go (hunk 2 above)
Comment (arsing) [inline tds.go]: Done, except for the ones where there was only one `else` branch.

File: tds.go (hunk 3 above)
Comment (arsing) [inline tds.go]: Done.

File: tds.go (hunk 1 above)
Comment (arsing) [inline tds.go]: Wait, I see what you meant now.

File: README.md [hunk 5]
```diff
@@ -46,11 +46,77 @@ Example:
...
+    db, err := sql.Open("mssql", "odbc:server=localhost;user id=sa;password={foo{bar}") // password is "foo{bar"
+    db, err := sql.Open("mssql", "odbc:server=localhost;user id=sa;password={foo}}bar}") // password is "foo}bar"
+    ```
+
+1. URL format with `sqlserver` protocol - `sqlserver://username:password@host/instance?param1=value&param2=value`
+
+    Example:
+
+    ```go
+    db, err := sql.Open("mssql", "sqlserver://someuser:foo%3A%2F%5C%21~%40;bar@somehost?connection+timeout=30") // password is ""foo:/\\!~@;bar""
```
Comment (kardianos) [inline README.md]: Please also include a more normal url string with fewer encodings and an instance like:
`sqlserver://dbuser:dbpass@hostname/instance?database=master&connection+timeout=30`

File: README.md [hunk 6]
```diff
@@ -46,11 +46,77 @@ Example:
...
+    ```
+
+1. ODBC format - `key=value` pairs separated by `;`. The string is prefixed with `odbc:`
+
+    This format allows characters like `;` within a value if the value is wrapped in `{}`. Leading and trailing whitespace is retained for wrapped values.
+
+    Example:
+
+    ```go
+    db, err := sql.Open("mssql", "odbc:server=localhost;user id=sa;password={foo;bar}") // password is "foo;bar"
```
Comment (kardianos) [inline README.md]: I think the more common question will also be to specify the instance name `server=localhost\SQLExpress;...` add example with that please.

Comment (arsing) [inline README.md]: Good point. I wasn't thinking of sqlservr on Windows :)

**Pull request merged**